### PR TITLE
Fix common dialog layout

### PR DESCRIPTION
**Pull request by MatkovIvan**
## Proposed Changes

- Extract common part from `PopupLayout` to `RootLayout`
- Port `DialogLayout` width logic from Android to iOS (stepped) and adopted to Desktop
- Add `usePlatformDefaultWidth` and `scrimColor` as experimental parameters to `DialogProperties`

## Testing

Test: run mpp sample, open max size common dialog, try to resize window

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform-core/pull/708#discussion_r1277706609

### CODE REVIEW

File: compose/ui/ui/src/notMobileMain/kotlin/androidx/compose/ui/window/Dialog.notMobile.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+// Adopted version from Android that doesn't use height for width calculation.
+private fun MeasureScope.preferredDialogWidth(constraints: Constraints): Int {
+    val width = constraints.maxWidth.toDp()
+    return when {
+        width >= 600.dp -> 580.dp
+        width >= 480.dp -> 440.dp
+        else -> 320.dp
+    }.roundToPx()
+}
```
Comment (m-sasha) [inline compose/ui/ui/src/notMobileMain/kotlin/androidx/compose/ui/window/Dialog.notMobile.kt]: I thought the decision was to make the sizing identical to mobile...?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt [hunk 2]
```diff
@@ -31,27 +33,43 @@ import androidx.compose.ui.input.key.type
...
+import androidx.compose.ui.util.fastMaxBy
 
 /**
  * The default scrim opacity.
- * See generated [androidx.compose.material3.tokens.ScrimTokens.ContainerOpacity] as reference.
- *
- * TODO: Provide a way to configure dialog shim color (maybe inside DialogProperties?)
  */
-private const val DefaultScrimOpacity = 0.32f
+private const val DefaultScrimOpacity = 0.6f
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:54]: Why is it 0.6 in the end?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt [hunk 3]
```diff
@@ -31,27 +33,43 @@ import androidx.compose.ui.input.key.type
...
+private val DefaultScrimColor = Color.Black.copy(alpha = DefaultScrimOpacity)
 
 @Immutable
-actual class DialogProperties actual constructor(
-    actual val dismissOnBackPress: Boolean,
-    actual val dismissOnClickOutside: Boolean
+actual class DialogProperties @ExperimentalComposeUiApi constructor(
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:70]: Need to document the additional parameters.
In fact, I think maybe we should document everything, even when the actual is exactly like the expect, to make it easier for the user and IDE to find the documentation.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt [hunk 4]
```diff
@@ -31,27 +33,43 @@ import androidx.compose.ui.input.key.type
...
+        dismissOnBackPress: Boolean,
+        dismissOnClickOutside: Boolean
+    ) : this(
+        dismissOnBackPress = dismissOnBackPress,
+        dismissOnClickOutside = dismissOnClickOutside,
+        usePlatformDefaultWidth = true,
+        scrimColor = DefaultScrimColor
+    )
+
     override fun equals(other: Any?): Boolean {
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:86]: Don't we need to update `equals` and `hashCode` now?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt [hunk 5]
```diff
@@ -102,24 +120,71 @@ actual fun Dialog(
...
+            measurePolicy = measurePolicy
+        )
+    }
+}
+
+@Composable
+private fun rememberDialogMeasurePolicy(
+    properties: DialogProperties,
+    onBoundsChanged: (IntRect) -> Unit
+) = remember(properties) {
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt]: Shouldn't `onBoundsChanged` be one of the keys?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt [hunk 6]
```diff
@@ -332,13 +347,83 @@ fun Popup(
...
+    }
+}
+
+@Composable
+private fun rememberPopupMeasurePolicy(
+    popupPositionProvider: PopupPositionProvider,
+    layoutDirection: LayoutDirection,
+    parentBounds: IntRect,
+    onBoundsChanged: (IntRect) -> Unit
+) = remember(popupPositionProvider, layoutDirection, parentBounds) {
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt]: Shouldn't `onBoundsChanged` be one of the keys?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/RootLayout.skiko.kt [hunk 7]
```diff
@@ -19,48 +19,30 @@ package androidx.compose.ui.window
...
 import androidx.compose.ui.platform.setContent
+import androidx.compose.ui.requireCurrent
 import androidx.compose.ui.unit.IntRect
-import androidx.compose.ui.unit.IntSize
 import androidx.compose.ui.unit.round
 
 @Composable
-internal fun PopupLayout(
-    popupPositionProvider: PopupPositionProvider,
+internal fun RootLayout(
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/RootLayout.skiko.kt:40]: Add at least a short documentation of what it's for.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt [hunk 8]
```diff
@@ -102,24 +120,71 @@ actual fun Dialog(
...
+    MeasurePolicy { measurables, constraints ->
```
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:175]: Maybe reuse this layout between dialog and popup (and use "positioner" again)?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt (hunk 2 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:54]: It's not material (so spec isn't relevant here). For compatibility with Android

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt (hunk 8 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Dialog.skiko.kt:175]: It's already done in separate branch/next PR

**Review by m-sasha — APPROVED** [review]

**Pull request merged**